### ISSUE TITLE
[🐛] Anisotropic filtering causes texture artifacts in true color

### ISSUE THREAD

**Issue by JakeSmarter** [luciusDXL/TheForceEngine#392]
Version: v1.09.540-264-g341db341+
OS: Linux
Distribution: Flathub

When “Anisotropic Filter Quality” is set to `1.0` in “True Color” mode you can see artifacts on texture edges where textures repeat. This happens when using the hardware GPU as well as Mesa 3D’s software renderer. This leads me to suspect that there is a bug in the anisotropic filter algorithm.
<TABLE><TR><TH>OpenGL<BR>Broadcom<BR>V3D 4.2<BR>true color</TH><TH>OpenGL<BR>Mesa<BR>llvmpipe (LLVM 17.0.6, 128 bits)<BR>true color</TH><TH>OpenGL<BR>Mesa<BR>llvmpipe (LLVM 17.0.6, 128 bits)<BR>8-bit classic</TH></TR>
<TR><TD><IMG SRC="https://github.com/luciusDXL/TheForceEngine/assets/15823872/7c35da32-a855-46b0-bd5e-453f7fca0a8e" ALT="OpenGL Broadcom V3D 4.2 true color"></TD><TD><IMG SRC="https://github.com/luciusDXL/TheForceEngine/assets/15823872/cfbb4c54-d316-49b6-a28a-096824508b27" ALT="OpenGL Mesa llvmpipe (LLVM 17.0.6, 128 bits) true color"></TD><TD><IMG SRC="https://github.com/luciusDXL/TheForceEngine/assets/15823872/7110f323-a699-4f6c-be39-c74dab49d562" ALT="OpenGL Mesa llvmpipe (LLVM 17.0.6, 128 bits) 8-bit classic"></TD></TR></TABLE>

**Comment by luciusDXL** [comment]
The algorithm is fine, the issue is that the setting is too high for the padding size and needs to be adjusted. This should be fixed in the next release.

**Comment by luciusDXL** [comment]
TFE uses GL_TEXTURE_MAX_LOD to modify the maximum mipmap/LOD value based on the anisotropic setting. It looks like that may not be supported in those screenshots, but is properly supported by GPUs tested on Windows. My guess is that this is due to GL_TEXTURE_MAX_LEVEL and GL_TEXTURE_MAX_LOD being different values. So you may need to turn the anisotropic value down to the minimum on your system (or disable mipmapping).

**Comment by mlauss2** [comment]
FWIW, I see a similar issue, but much less pronounced, also on Adreno/WindowsARM64:

![Screenshot 2024-08-14 214610](https://github.com/user-attachments/assets/372fd80e-9e28-4469-8314-c4dd1a489bfc)

Since the Adreno Windows OpenGL driver is based on Mesa (mesa-to-d3d12), this *might* be an issue with mesa.
BUT, I *don't* see this with the mesa drivers on radeon, intel or nvidia.

(the adreno has other problems, like this clipping issue:
![Screenshot 2024-08-14 214538](https://github.com/user-attachments/assets/1cedecd0-f0c4-4038-a494-bbbf5392e0da)
)

**Comment by luciusDXL** [comment]
> However, on one hand I can point to a handful of other apps that work flawlessly with anisotropic filtering using both Mesa 3D drivers.

This isn't surprising. The issue with TFE is that it uses a TextureArray of texture atlases, which means that the maximum lod is *very* important to avoid reading outside of the current tile during sampling. If drivers don't handle this properly, you get these issues. Most apps you are testing against are using anisotropic filtering with full-sized textures.

I can possibly add an option to add a mip bias when anisotropic filtering is enabled (which would be bad for devices that support this properly since it will add more aliasing, but better than no mipmapping at all I think - but makes sense as an option). My biggest issue is that I don't have hardware setup that exhibits this issue, so it is hard for me to test any such changes.

**Comment by luciusDXL** [comment]
If you are running on hardware with limited GPU power, you are probably better off running the software renderer. The GPU Renderer is designed for semi-modern GPU hardware and pushes more work to the GPU - thus having heavier shaders (GPU expansion, texture projection, etc.). That is why the GPU renderer was designed for OpenGL 3.3+ hardware - with the expectation that lower end systems would use software with modest resolutions.

In other words, on modern systems I get much *better* performance than eDuke32 with similar complexity (thousands of fps versus hundreds) - with an eye toward future more complex mods. But by optimizing for newer hardware, it means the shaders and GPU is typically too heavy for lower-end devices.
